File: heimdall/src/main/java/de/rheinfabrik/oauth2/OAuth2AccessTokenStorage.java
package de.rheinfabrik.oauth2;

import rx.Observable;

/**
 * Interface used to define how to store and retrieve a stored access token.
 *
 * @param <TAccessToken> The access token type.
 */
public interface OAuth2AccessTokenStorage<TAccessToken extends OAuth2AccessToken> {

    // Public API

    /**
     * Checks whether there is or is not an access token.
     *
     * @return - An Observable emitting a boolean representing whether there is or is not an access token.
     */
    public Observable<Boolean> hasAccessToken();

    /**
     * Queries the stored access token.
     *
     * @return - An Observable emitting the stored access token.
     */
    public Observable<TAccessToken> getStoredAccessToken();

    /**
     * Stores the given access token.
     *
     * @param token The access token which will be stored.
     */
    public void storeAccessToken(TAccessToken token);
}

File: library/src/main/java/de/rheinfabrik/heimdall2/coroutines/OAuth2AccessTokenStorage.kt
package de.rheinfabrik.heimdall2.coroutines

import de.rheinfabrik.heimdall2.accesstoken.OAuth2AccessToken

/**
 * Interface used to define how to store and retrieve a stored access token.
 *
 * @param <TAccessToken> The access token type.
 */
interface OAuth2AccessTokenStorage {

    // Public API

    /**
     * Queries the stored access token.
     *
     * @return - An stored access token.
     */
    suspend fun getStoredOAuth2AccessToken(): OAuth2AccessToken

    /**
     * Stores the given access token.
     *
     * @param token The access token which will be stored.
     */
    suspend fun storeOAuth2AccessToken(token: OAuth2AccessToken)

    /**
     * Checks whether there is or is not an access token
     *
     * @return - An Boolean based on whether there is or is not an access token.
     */
    suspend fun hasOAuth2AccessToken(): Boolean

    /**
     * Removes the stored access token.
     */
    suspend fun removeOAuth2AccessToken()
}
